lib/pbench/cli/server/report.py

```diff
@@ -0,0 +1,403 @@
+from collections import defaultdict
+import datetime
+import re
+from typing import Optional
+
+import click
+import humanize
+from sqlalchemy import inspect, select, text
+
+from pbench.cli import pass_cli_context
+from pbench.cli.server import config_setup
+from pbench.cli.server.options import common_options
+from pbench.client.types import Dataset
+from pbench.common.logger import get_pbench_logger
+from pbench.server import BadConfig
+from pbench.server.cache_manager import CacheManager
+from pbench.server.database.database import Database
+from pbench.server.database.models.datasets import Metadata
+from pbench.server.database.models.index_map import IndexMap
+
+
+class Detail:
+    """Encapsulate generation of additional diagnostics"""
+
+    def __init__(self, detail: bool):
+        self.detail = detail
+
+    def __bool__(self) -> bool:
+        return self.detail
+
+    def write(self, message: str):
+        if self.detail:
+            click.echo(f"|| {message}")
+
+
+class Verify:
+    """Encapsulate -v status messages."""
+
+    def __init__(self, verify: bool):
+        self.verify = verify
+
+    def __bool__(self) -> bool:
+        return self.verify
+
+    def status(self, message: str):
+        if self.verify:
+            ts = datetime.datetime.now()
+            click.echo(f"({ts:%H:%M:%S}) {message}")
+
+
+class Watch:
+    """Encapsulate a periodic status check.
+
+    Discovery (especially for cache and backup) can take a long time, so we
+    centralize a periodic update notice mechanism.
+    """
+
+    def __init__(self, interval: float):
+        self.interval = datetime.timedelta(seconds=interval) if interval else None
+        self.start = datetime.datetime.now()
+        self.last = self.start
+
+    def update(self, status: str):
+        now = datetime.datetime.now()
+        if self.interval and now > self.last + self.interval:
+            self.last = now
+            delta = now - self.start
+            hours, remainder = divmod(delta.seconds, 3600)
+            minutes, seconds = divmod(remainder, 60)
+            click.echo(f"[{hours:02d}:{minutes:02d}:{seconds:02d}] {status}")
+
+
+detailer: Optional[Detail] = None
+watcher: Optional[Watch] = None
+verifier: Optional[Verify] = None
+
+
+def report_archive(tree: CacheManager):
+    """Report archive statistics.
+
+    Args:
+        tree: a cache instance
+    """
+
+    tarball_count = len(tree.datasets)
+    tarball_size = 0
+    smallest_tarball = 0
+    smallest_tarball_name = None
+    biggest_tarball = 0
+    biggest_tarball_name = None
+
+    for tarball in tree.datasets.values():
+        watcher.update(f"({tarball_count}) archive {tarball.name}")
+        size = tarball.tarball_path.stat().st_size
+        tarball_size += size
+        if not smallest_tarball or size < smallest_tarball:
+            smallest_tarball = size
```

Comment:
I would routinely do that for C where there's a clear `MAXINT`; but Python doesn't really even have that concept, and if it exposes a filesystem maximum file size, I haven't encountered an interface to get it. I hate choosing an arbitrary value, and this more flexible alternative doesn't add much complexity.

Reply:
As you say, this is fine.  However, I would think an arbitrary value would suffice (say, 2^63?):  I don't think that there are any filesystems of that size yet, and, even if there were, it seems unlikely in the extreme that we would have any single file approaching that size (let alone the hardware to support it).  And, given that we have a large number of tarballs and a human waiting for the result, skipping otherwise unneeded checks might be a help.